### DISCUSSION TITLE
Question on building efu for OTA update

### DISCUSSION THREAD

**Discussion by pakeonoahu** [forkineye/ESPixelStick#499] [General]
Hi,
     I have no issue flashing the firmware and spiffs using platformio.  I want to test the OTA firmware update, looks like the ESPSFlashTool uses *-app.bin, *-bootloader.bin, *-partition.bin  and fs folder to build .efu.  The platformio built ended up with firmware.bin, partitons.bin and spiffs.bin, how do I build .efu file with those? how do I match/rename those files? 

thanks

**Comment by MartinMueller2003** [comment]
First we do hot use spiffs so the build is not using the latest version of our pio scripts. Yes we rename the firmware and partition files to conform to the names in the flash tool json file. You can put any file names you want in that file. The bootloader file is copied from the board tools directory. 

Since you are using PIO, you can drop the attached file into your ESPixelStick/.scripts directory and the files will be generated for you and dropped into the ESPixelStick/dist/firmware directory at the end of the build. You will need to remove the .txt from the end of the file name to get it to work.
[CopyTargets.py.txt](https://github.com/forkineye/ESPixelStick/files/8402407/CopyTargets.py.txt)

**Comment by pakeonoahu** [comment]
Thank you for the .py file, but for some reason, when I build in PIO, it doesn't execute it, I installed Python 3.10.4 as well.   I couldn't run it alone since it needs env from PIO(NameError: name 'Import' is not defined).  I ended up copying the firmware to -app.bin and partitions.bin to the flasher directory of a beta release, also copy the new data/www files into the fs and overwrite the firmware.json as well, able to build the .efu and flash it using OTA.  It works without any issues. 
Thanks.

**Reply by MartinMueller2003** [reply]
My apologies. My instructions missed a step. Add the lines below to the platformio_user.ini file and you should be good to go.

[esp8266]
extra_scripts = ${env.extra_scripts}
                post:.scripts/CopyTargets.py

[esp32]
extra_scripts = ${env.extra_scripts}
                .scripts/replace_fs.py
                post:.scripts/CopyTargets.py

**Comment by pakeonoahu** [comment]
Hi, I tried to add the post script on the [esp32] but didn't work.

Error: Invalid 'platformio_user.ini' (project configuration file): 'Source contains parsing errors: 'platformio_user.ini'
        [line 43]: '.scripts/replace_fs.py\n''

removing it like the [esp8266], no errors but doesn't run the script either.

**Comment by MartinMueller2003** [comment]
Here are my my platformio files. They work without errors
[pio.zip](https://github.com/forkineye/ESPixelStick/files/8409065/pio.zip)

And here is the contents of my scripts directory (I have a few more than the releases).
[scripts.zip](https://github.com/forkineye/ESPixelStick/files/8409092/scripts.zip)

**Comment by pakeonoahu** [comment]
Thank you very much! it works perfectly with your files, I noticed some diff on the platformio.ini and platformio_user.ini between your and the beta releases.

**Reply by MartinMueller2003** [reply]
yes i have been adding platforms and integrating newer versions of libraries.

**Comment by MartinMueller2003** [comment]
You can build EFU files at: https://espixelstickwebflasher.from-ct.com:5000/
